propagation.go

```diff
@@ -41,71 +45,115 @@ var (
 type BuiltinFormat byte
 
 const (
-	// SplitBinary encodes the Span in a SplitBinaryCarrier instance.
+	// Binary encodes the Span for propagation as opaque binary data.
+	//
+	// For Tracer.Inject(): the carrier must be an `io.Writer`.
 	//
-	// The `carrier` for injection and extraction must be a
-	// `*SplitBinaryCarrier` instance.
-	SplitBinary BuiltinFormat = iota
+	// For Tracer.Join(): the carrier must be an `io.Reader`.
+	Binary BuiltinFormat = iota
 
-	// SplitText encodes the Span in a SplitTextCarrier instance.
+	// TextMap encodes the Span in a TextMapCarrier instance.
+	//
+	// For Tracer.Inject(): the carrier must be a `TextMapWriter`.
+	//
+	// For Tracer.Join(): the carrier must be a `TextMapReader`.
+	//
+	// See HTTPHeaderTextMapCarrier for an implementation of both TextMapWriter
+	// and TextMapReader that defers to an http.Header instance for storage.
+	// For example, Inject():
+	//
+	//    carrier := HTTPHeaderTextMapCarrier(httpReq.Header)
+	//    err := span.Tracer().Inject(span, TextMap, carrier)
 	//
-	// The `carrier` for injection and extraction must be a `*SplitTextCarrier`
-	// instance.
+	// Or Join():
+	//
+	//    carrier := HTTPHeaderTextMapCarrier(httpReq.Header)
+	//    span, err := tracer.Join("opName", TextMap, carrier)
+	//
+	TextMap
+
+	// SplitBinary is DEPRECATED
+	SplitBinary
+	// SplitText is DEPRECATED
 	SplitText
+)
+
+// TextMapWriter is the Inject() carrier for the TextMap builtin format. With
+// it, the caller can encode a Span for propagation as entries in a multimap of
+// unicode strings.
+type TextMapWriter interface {
+	// Add a key:value pair to the carrier. Multiple values may be added for a
+	// single (repeated) key.
+	Add(key, val string)
+}
 
-	// GoHTTPHeader encodes the Span into a Go http.Header instance (both the
-	// tracer state and any baggage).
+// TextMapReader is the Join() carrier for the TextMap builtin format. With it,
+// the caller can decode a propagated Span as entries in a multimap of unicode
+// strings.
+type TextMapReader interface {
+	// ReadAll returns TextMap contents via repeated calls to the `handler`
+	// function. If any call to `handler` returns a non-nil error, ReadAll
+	// terminates and returns that error.
 	//
-	// The `carrier` for both injection and extraction must be an http.Header
-	// instance.
-	GoHTTPHeader
-)
+	// NOTE: A single `key` may appear in multiple calls to `handler` for a
+	// single `ReadAllEntries` invocation.
```

Comment:
My concern is to simplify the implementation of the carrier by RPC owners. If RPC supports application headers as a plain map, then multi-map semantics of the Writer makes it difficult to implement. And conversely, if RPC supports multi-map headers, then requiring plain map in the Reader makes it difficult to implement. So rather than going with additional prescriptions like "concatenate values with semicolon" or "call handler only on the first value", I'd prefer an API that imposes minimal restrictions in each case, even if it makes it asymmetrical.

What do you think?